src/semidiscretization/semidiscretization_coupled.jl

```diff
@@ -137,23 +140,36 @@ end
     @view u_ode[semi.u_indices[index]]
 end
 
+# Same as `foreach(enumerate(something))`, but without allocations.
+#
+# Note that compile times may increase if this is used with big tuples.
+@inline foreach_enumerate(func, collection) = foreach_enumerate(func, collection, 1)
+@inline foreach_enumerate(func, collection::Tuple{}, index) = nothing
+
+@inline function foreach_enumerate(func, collection, index)
+    element = first(collection)
+    remaining_collection = Base.tail(collection)
+
+    func((index, element))
+
+    # Process remaining collection
+    foreach_enumerate(func, remaining_collection, index + 1)
+end
+
 function rhs!(du_ode, u_ode, semi::SemidiscretizationCoupled, t)
     @unpack u_indices = semi
 
     time_start = time_ns()
 
-    @trixi_timeit timer() "copy to coupled boundaries" begin
-        for semi_ in semi.semis
-            copy_to_coupled_boundary!(semi_.boundary_conditions, u_ode, semi)
-        end
+    foreach(semi.semis) do semi_
+        copy_to_coupled_boundary!(semi_.boundary_conditions, u_ode, semi, semi_)
     end
 
     # Call rhs! for each semidiscretization
-    for i in eachsystem(semi)
+    foreach_enumerate(semi.semis) do (i, semi_)
         u_loc = get_system_u_ode(u_ode, i, semi)
         du_loc = get_system_u_ode(du_ode, i, semi)
-
-        @trixi_timeit timer() "system #$i" rhs!(du_loc, u_loc, semi.semis[i], t)
```

Comment:
Why was the timer removed? Was it added in a different location?

Reply:
Same reason as for the other timing bit. This, on the other hand, does lead to extra memory allocations. However, if we put the timing around the loop there is no extra allocation.

Reply:
If this is possible without allocations, great, if not, we'll just leave it as it is

Reply:
By putting the timer around the loop there are no extra allocations.

Reply:
It's already implemented in the latest commit.